gh-138045: Fix importing numpy in legacy subinterpreters by skipping switch to main interpreter when not needed

jan-hasse opened the pull request:
(singlephase == NULL) check in update_global_state_for_extension could be removed because it's only called with pointers to stack objects.

I've tried creating a minimal working reproducer instead of importing numpy with the example from https://github.com/python/cpython/issues/138045#issuecomment-3216214055 but didn't manage to integrate that into Python's test suite. Maybe someone else has more knowledge on how to create a simple test that reproduce the numpy situation.

The fix can be backported down to 3.13.

<!-- gh-issue-number: gh-138045 -->
* Issue: gh-138045
<!-- /gh-issue-number -->

ZeroIntensity reviewed (commented):
Please add a news entry, and ideally a test case. This seems difficult to test, though. You could look into adding something to `_testsinglephase`.

jan-hasse committed:
gh-138045: Fix importing numpy in legacy subinterpreters by skipping switch to main interpreter when not needed

(singlephase == NULL) check in update_global_state_for_extension could be removed because it's only called with pointers to stack objects.
